File: December 2024 Leetcode Solution/December-08.cpp
class Solution {
public:
    int maxTwoEvents(vector<vector<int>>& events) {
        sort(events.begin(), events.end());
        int n = events.size();
        vector<int> maxValues(n);
        maxValues[n - 1] = events[n - 1][2];

        for (int i = n - 2; i >= 0; --i) {
            maxValues[i] = max(maxValues[i + 1], events[i][2]);
        }

        vector<int> startTimes;
        for (const auto& event : events) {
            startTimes.push_back(event[0]);
        }

        int maxResult = 0;
        for (const auto& event : events) {
            int idx = upper_bound(startTimes.begin(), startTimes.end(), event[1]) - startTimes.begin();
            int value = event[2];
            if (idx < n) value += maxValues[idx];
            maxResult = max(maxResult, value);
        }

        return maxResult;
    }
};

File: December 2024 Leetcode Solution/December-08.py
from bisect import bisect_right
from typing import List

class Solution:
    def maxTwoEvents(self, events: List[List[int]]) -> int:
        events.sort()
        n = len(events)
        max_values = [0] * n
        max_values[-1] = events[-1][2]

        for i in range(n - 2, -1, -1):
            max_values[i] = max(max_values[i + 1], events[i][2])

        start_times = [event[0] for event in events]
        max_result = 0

        for start, end, value in events:
            idx = bisect_right(start_times, end)
            if idx < n:
                value += max_values[idx]
            max_result = max(max_result, value)

        return max_result
